### DISCUSSION TITLE
Iterate over string and pair in a JSON file.

### DISCUSSION THREAD

**Discussion by cxxboy** [nlohmann/json#2549] [Q&A] [answered]
Suppose my .json file looks something like this:
```json
{
	"/post_one": {
		"user_one": "sakfjsaljfjoijfoisadjfio@#42asdf",
		"user_two": "dfasfjlsjldkjfl;ksjadflkj@#42asdf"
	},
	"/post_two": {
		"user_foo": "sakfjsaljfjoijfoisadjfio@@$2asdf",
		"user_bar": "sakfjsaljfjoijfoisadjfio@#%#5324"
	}
}
```
I will be parsing this structure into an 
`std::unordered_map<std::string, std::vector<std::pair<std::string, std::string>>>`
for my usercase. But, How can I iterate over the `/post_one` and `/post_two` (which are keys) and parse the two items within that structure into strings so that I can first iterate over one by one (`/endpoints`) and inside that, I can again iterate over the structure and fill-in/populate my Map? Is there any way i can do this? I tried looking into`/doc` but cannot idk how I can something/some value whose key I may not know beforehand.

**Comment by nlohmann** [comment, marked as answer]
The [`items()`](https://json.nlohmann.me/api/basic_json/items/) function should do the job.

Example:

```cpp
#include "json.hpp"

using json = nlohmann::json;

int main()
{
    json j = R"({
    "/post_one": {
        "user_one": "sakfjsaljfjoijfoisadjfio@#42asdf",
        "user_two": "dfasfjlsjldkjfl;ksjadflkj@#42asdf"
    },
    "/post_two": {
        "user_foo": "sakfjsaljfjoijfoisadjfio@@$2asdf",
        "user_bar": "sakfjsaljfjoijfoisadjfio@#%#5324"
    }
})"_json;
    
    std::unordered_map<std::string, std::vector<std::pair<std::string, std::string>>> m;
    
    for (const auto & outer : j.items())
    {
        for (const auto & inner : outer.value().items())
        {
            m[outer.key()].emplace_back(inner.key(), inner.value());
        }
    }
}
```

**Comment by cxxboy** [comment]
@nlohmann OK, I marked it too early. Looks like this solution Does Not work for me.
Here is my complete code where I will be parsing the JSON from a "external file"
```c++
	std::ifstream cred_stream{JSON_FILE_NAME};
	std::string file_data;
	if(cred_stream.is_open()){
		std::string line;
		while(std::getline(cred_stream, line)){
			file_data.append(line);
		}
	}else{
	 	perror("Unable to open API auth file for REST services\n");
		::exit(1);
	}
	cred_stream.close();
	nlohmann::json json_reader(file_data);
	for(const auto& endpoints : json_reader.items()){
		for(const auto& credentials : endpoints.value().items()){
			std::cout << credentials.value() << std::endl;
			this->_endpoint_cred_map[endpoints.key()].emplace_back(credentials.key(), credentials.value());
		}
	}
```
But, when I try to print this out, the whole JSON file is dumped into `credentials.value()`, so it prints :-
`"{\t\"/post_one\": {\t\t\"user_one\": \"passwordHash1!@\",\t\t\"user_two\": \"passwordHash2#$\"\t},\t\"/post_two\": {\t\t\"user_foo\": \"passwordHash3$%\",\t\t\"user_bar\": \"passwordHash4%^*\"\t}}"`
Which is not what I wanted.

**Reply by nlohmann** [reply]
Line

```cpp
nlohmann::json json_reader(file_data);
```

does not parse the content of `file_data`, but creates a JSON value consisting of the string `file_data`.

Try replacing it with

```cpp
auto json_reader  = nlohmann::json::parse(file_data);
```

**Reply by cxxboy** [reply]
This works perfectly. Thank you for your fast response :-)
